src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -0,0 +1,640 @@
+        /// </summary>
+        /// <remarks>
+        /// The names and the order of the parameters are the same as for the corresponding Q# operation.
+        /// Class implementing <see cref="IQuantumProcessor"/> interface can return any class derived from <see cref="Result"/>.
+        /// </remarks>
+        /// <param name="qubits">Qubits to which the gate should be applied.</param>
+        /// <param name="bases">Array of single-qubit Pauli values describing multi-qubit Pauli observable.</param>
+        /// <returns> Zero if the +1 eigenvalue is observed, and One if the -1 eigenvalue is observed.</returns>
+        Result Measure(IQArray<Pauli> bases, IQArray<Qubit> qubits);
+
+        /// <summary>
+        /// Called when <a href="https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.reset">Microsoft.Quantum.Intrinsic.Reset</a> is called in Q#.
+        /// When this is invoked, it is expected that the <paramref name="qubit"/> is measured and ensured to be in the |0⟩ state such that it can be safely released.
+        /// </summary>
+        /// <remarks>
+        /// The names and the order of the parameters are the same as for the corresponding Q# operation.
+        /// </remarks>
+        /// <param name="qubit">Qubit to which the gate should be applied.</param>
+        void Reset(Qubit qubit);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when a conditional statement on measurement results is invoked.
+        /// </summary>
+        /// <param name="measurementResults">The actual results of the measurements of a number of qubits upon which branching is to be performed.</param>
+        /// <param name="resultsValues">The expected values of results of the measurements of these qubits.</param>
+        /// <returns> A value representing this conditional statement and encoding the result of condition.</returns>
+        /// <remarks>
+        /// A typical implementation will compare all <paramref name="measurementResults"/> to <paramref name="resultsValues"/> and return the result of this comparison.
+        /// </remarks>
+        int StartConditionalStatement(IQArray<Result> measurementResults, IQArray<Result> resultsValues);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when a conditional statement on a measurement result is invoked.
+        /// </summary>
+        /// <param name="measurementResult">The actual result of the measurement of a qubit upon which branching is to be performed.</param>
+        /// <param name="resultValue">The expected value of result of the measurement of this qubit.</param>
+        /// <returns> A value representing this conditional statement and encoding the result of condition.</returns>
```

Comment:
Maybe it is worth mentioning that the return value will be passed as a `statement` argument to `RunThenClause`, `RepeatThenClause`, `RunElseClause`, `RepeatElseClause` and `EndConditionalStatement`. I wonder where is the good place to desribe the whole mechanism together.

Reply:
Minor: Is using `long` instead of `int` is slightly more future-proof?